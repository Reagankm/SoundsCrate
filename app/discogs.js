/* @flow */
import { secureFetch } from './lib/oauth';
import conf from './conf';
import cache from './cache';

import type {
  Folder,
  RecordsPage,
  FolderPage,
  FolderPageReleases
} from './discogsTypes';

export async function getCollectionFolder(folderId: string, pageNumber: number): Promise<RecordsPage> {
  const cachedFolder = cache.get(folderId, pageNumber);
  if (cachedFolder) {
    const { totalPages, records } = cachedFolder;
    return { totalPages, records };
  } else {
     const { totalPages, releases } = await getCollectionFolderPage(folderId, pageNumber);
     const records = releases.map(getReleaseCoverImage);
     const entry = { pageNumber, records };
     cache.update({
       folderId,
       entry,
       totalPages
     });

     return { totalPages, records };
  }
}

export async function getThumbsInFolder(folderId: string, numberOfThumbs: number = 4): Promise<*> {
  const { totalPages, releases } = await getCollectionFolderPage(folderId, 1, numberOfThumbs);
  return releases.map(getReleaseCoverImage);
}

export async function getCollectionFolders(): Promise<Array<Folder>> {
  const {
    discogs: {
      api_url: apiUrl,
      endpoints,
    }
  } = conf;
  const { username } = await getIdentity();
  const foldersCollectionUrl = endpoints.folders.replace('{username}', username);
  const foldersUrl = `${apiUrl}${foldersCollectionUrl}`;
  const foldersResponse = await secureFetch(foldersUrl);
  const { folders } = await getJson(foldersResponse);

  return folders.map(folder => ({
    id: folder.id,
    name: folder.name,
    count: folder.count,
  }));
}

async function getCollectionFolderPage(
  folderId: string,
  pageNumber: number,
  recordsPerPage?: number): Promise<FolderPage> {
  const {
    discogs: {
      api_url: apiUrl,
      endpoints,
      records_per_page,
    }
  } = conf;
  const { username } = await getIdentity();
  const perPage = recordsPerPage || records_per_page;
  const userFolderUrl = endpoints.folder.replace('{username}', username).replace('{id}', folderId.toString());
  const folderUrl = `${apiUrl}${userFolderUrl}?sort=artist&per_page=${perPage}&page=${pageNumber}`;
  const folderResponse = await secureFetch(folderUrl);
  const {
    releases,
    pagination: { pages: totalPages }
  } = await getJson(folderResponse);

  return { totalPages, releases };
}

async function getIdentity(): Promise<*> {
  const {
    discogs: {
      api_url: apiUrl,
      endpoints,
    }
  } = conf;
  const response = await secureFetch(`${apiUrl}${endpoints.identity}`);
  return await getJson(response);
}

function getJson(response: Response): Promise<*> {
  return response.json();
}

function getReleaseCoverImage(release: FolderPageReleases): string {
  return release.basic_information.cover_image;
}
