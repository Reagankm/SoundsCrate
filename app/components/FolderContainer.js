/* @flow */
import React, { Component } from 'react';
import {
  Dimensions,
  Image,
  StyleSheet,
  Text,
  TouchableHighlight,
  View,
} from 'react-native';
import record from './assets/record.png';
import { getCollectionFolder } from '../discogs';
import Folder from './Folder';
import LoadingIndicator from './LoadingIndicator';

type State = {
  records: Array<string>,
  totalPages: number,
  currentPage: number,
};

export default class FolderContainer extends Component {
  state: State = { records: [], currentPage: 1, totalPages: 1 };

  componentWillMount() {
    this.getCollectionFolder(this.state.currentPage);
  }

  render() {
    const { records } = this.state;
    return (
      <View style={styles.root}>
        {!Boolean(records.length) && <LoadingIndicator/>}
        {Boolean(records.length) && <Folder
          records={records}
          onScrollEnd={() => this.onFolderScrollEnd()}/>}
      </View>
    );
  }

  getCollectionFolder(pageNumber: number) {
    this.getFolder(pageNumber);
  }

  onFolderScrollEnd() {
    const { currentPage, totalPages } = this.state;
    if (currentPage < totalPages) {
      this.getCollectionFolder(this.state.currentPage + 1);
    } else {
      console.warn('No more records to load!');
    }
  }

  async getFolder(pageNumber: number) {
    const { state } = this.props.navigation;
    const recordsPage = await getCollectionFolder(
      state.params.id,
      pageNumber,
    );
    this.setState({
      currentPage: pageNumber,
      records: this.state.records.concat(recordsPage.records),
      totalPages: recordsPage.totalPages,
    });
  }
}

const styles = StyleSheet.create({
  root: { flex: 1 },
});
