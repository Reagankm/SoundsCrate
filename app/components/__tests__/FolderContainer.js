/* @flow */
import React from 'react';
import renderer from 'react-test-renderer';
import FolderContainer from '../FolderContainer';
import { getCollectionFolder } from '../../discogs';

jest.mock('../Folder', () => 'Folder');
jest.mock('../../discogs', () => ({
  __esModule: true,
  getCollectionFolder: jest.fn()
}));

describe('app/components/FolderContainer', () => {
  it('should render and call discogs.getCollectionFolder', async () => {
    const tree = await renderer.create(
      <FolderContainer navigation={{ state: { params: { id: 1 } } }} />);
    expect(getCollectionFolder).toHaveBeenCalledWith(1, 1);
    expect(tree).toMatchSnapshot();
  });
});
