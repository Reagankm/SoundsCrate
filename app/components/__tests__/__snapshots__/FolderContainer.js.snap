// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/components/FolderContainer should render and call discogs.getCollectionFolder 1`] = `
<Folder
  isLoading={true}
  onScrollEnd={[Function]}
  records={Array []}
  totalPages={1}
/>
`;
